bluetooth: Add bt_disable(), hci.close() and its implementation on stm32wb

erwango opened the pull request:
In order to allow shuting down C2 core on stm32wb SoCs, an application callable function is required to properly executing the required procedure from HCI IPM driver (see #41944).

Propose a `bt_disable()` function, together with a hci `close()` implementation to achieve this.

STM32 dedicated sample has been updated to demonstrate:
- bt_disable() successful behavior
- bt_enable() could be successfully applied after bt_disable() was performed

Likely deserves some polishing:
- bt dedicated tests update
- API documentation clarification
- Check on the correctness of changes applied to BT_DEV_ENABLE/BT_DEV_DISABLE flags

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/41944 (Instead of https://github.com/zephyrproject-rtos/zephyr/pull/42309)

Based on:
- ~~#42274~~ (merged)
- ~~#42368~~ (merged)

erwango committed:
include/drivers/bluetooth: hci_driver: Add close function

Add a close api function to bt_hci_driver structure.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
subsys/bluetooth/host: hci_core: Add a bt_disable command

Add a bt_disable function. This will abort existing threads and
close the HCI transport.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/bluetooth/hci: stm32wb: Move RX thread creation to open

In order to enable RX thread abortion on close, move creation
to from init to open.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/bluetooth/hci: stm32wb: Implement close api

Add treatment for close api.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
soc/arm/st_stm32: stm32wb: Remove stack reset from shutdown

BLE stack reset should now be done in hci close function.
Remove from shutdown procedure.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
samples/boards/stm32: stm32_ble: Use bt_disable to reset the BLE stack

Application is now supposed to explicitly shut down the ble stack
before shutdown.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/bluetooth: stm32wb: Allow sequential C2 start

Set up a c2_reset procedure in order to allow sequential
open/close/open calls and keep c2_reset done at init (required
for flash access).
Move reinit out of the reset procedure, so flash could be
accessed after bt_disable().

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
samples/boards/stm32: stm32_ble: Add bt_disable/bt_enable sequence

Demonstrate bt_disable/bt_enable.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
^^ @r2r0 in place of https://github.com/zephyrproject-rtos/zephyr/pull/42309

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3641,6 +3641,46 @@ int bt_enable(bt_ready_cb_t cb)
+		return -EALREADY;
+	}
+
+	err = bt_dev.drv->close();
```
You can't assume that drivers implement a `close()` callback. This is important for backwards compatibility, and also if a `close()` callback doesn't make sense for some drivers (possibly most of them). So you should probably have a check right after the `if (!bt_dev.drv)` check:
```c
if (!bt_dev.drv->close) {
	return -ENOTSUP;
}
```

erwango replied:
Nice, sure. Added.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3641,6 +3641,50 @@ int bt_enable(bt_ready_cb_t cb)
+		return -EALREADY;
+	}
+
+	if (!bt_dev.drv->close) {
+		return -ENOTSUP;
+	}
```
You need to have this check before making any state modifications, i.e. you need to move it before clearing `BT_DEV_ENABLE`.

erwango replied:
Right, fixed.

r2r0 commented:
I see "Bluetooth disable failed (err -120)" message as the result of the latest `bt_disable` call.
After adding `atomic_clear_bit(bt_dev.flags, BT_DEV_DISABLE);` at the begining of `bt_init` everything looks Ok.

erwango commented:
> I see "Bluetooth disable failed (err -120)" message as the result of the last `bt_disable` call. After adding `atomic_clear_bit(bt_dev.flags, BT_DEV_DISABLE);` at the begining of `bt_init` everything looks Ok.

Thanks @r2r0. I had it fixed originally, but apparently I somehow lost it during a rebase. Setting it back (in `bt_enable` to keep symetry with mnagement of `BT_DEV_ENABLE` flag).

erwango commented:
Rebased after https://github.com/zephyrproject-rtos/zephyr/pull/42274 was merged.

erwango commented:
Rebased, following merge of https://github.com/zephyrproject-rtos/zephyr/pull/42368

ceolin commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3641,6 +3643,50 @@ int bt_enable(bt_ready_cb_t cb)
+
+	bt_monitor_send(BT_MONITOR_CLOSE_INDEX, NULL, 0);
+
+	atomic_set_bit(bt_dev.flags, BT_DEV_DISABLE);
```
wasn't is already set in line 3661 ? Probably you should just check there and set the bit here ...

erwango replied:
Thanks, fixed

ceolin commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3641,6 +3643,50 @@ int bt_enable(bt_ready_cb_t cb)
+		return -ENOTSUP;
+	}
+
+	atomic_clear_bit(bt_dev.flags, BT_DEV_ENABLE);
```
shouldn't it be done in the end of this function ? Otherwise another thread can call `bt_enable()` while this one is still disabling it ...

erwango replied:
Makes sense, moved.

ceolin commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -29,6 +29,7 @@ enum {
 /* bt_dev flags: the flags defined here represent BT controller state */
 enum {
 	BT_DEV_ENABLE,
+	BT_DEV_DISABLE,
```
Is this  necessary ? Couldn't enable / disable be controlled by one single bit ?

erwango replied:
Having a DISABLE bit prevents hci.close() reentry which is something required as this can involve complex procedure like communication with another core.

ceolin replied:
:+1:

erwango commented:
@jhedberg, @ceolin  Thanks for the review, I hope I addressed all the comments.

Btw, I'm getting an error in Bluetooth tests, it doesn't seems linked and I'm not sure what to do about it
`tests/bluetooth/bsim_bt/bsim_test_mesh/tests_scripts/provision/pb_adv_reprovision.sh FAILED`

ceolin commented on samples/boards/stm32/power_mgmt/stm32wb_ble/src/main.c:
```diff
@@ -116,7 +124,27 @@ void main(void)
+
+	printk("BLE restart\n");
+	/* Initialize the Bluetooth Subsystem */
+	err = bt_enable(bt_ready);
```
nitpick. It should check `err` value.

erwango commented:
@ceolin Your review were dismissed when fixing the remark. Would you mind approving again ?

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3627,6 +3629,49 @@ int bt_enable(bt_ready_cb_t cb)
+
+	bt_monitor_send(BT_MONITOR_CLOSE_INDEX, NULL, 0);
+
+	atomic_clear_bit(bt_dev.flags, BT_DEV_ENABLE);
```
I just took a look at #41424 and it's adding an API For checking BT_DEV_READY instead of BT_DEV_ENABLE. How do these relate? It seems that either this or that PR should be changed (ping @JordanYates), since if we don't clear BT_DEV_READY then the new `bool bt_dev_ready(void);` API that @JordanYates adds wouldn't work with `bt_disable()` as added here.

JordanYates replied:
Clearing `BT_DEV_READY` should definitely be done. Most Bluetooth API's are checking whether `BT_DEV_READY` is set, not `BT_DEV_ENABLE`.

erwango replied:
@jhedberg I've added BT_DEV_READY clearing.

erwango replied:
@jhedberg I've fixed the point.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3627,6 +3629,53 @@ int bt_enable(bt_ready_cb_t cb)
+	}
+
+	/* Clear BT_DEV_READY before disabling HCI link */
+	atomic_clear_bit(bt_dev.flags, BT_DEV_READY);
+
+	err = bt_dev.drv->close();
+	if (err) {
+		BT_ERR("HCI driver close failed (%d)", err);
+		return err;
+	}
```
Won't this leave the stack in an inconsistent and unrecoverable state? You've cleared the `BT_DEV_READY` flag, but not the `BT_DEV_ENABLE` flag. Should the `BT_DEV_READY` be re-enabled in this error branch?

erwango replied:
Indeed, this is probably safer to set it back. Done

erwango commented:
> @erwango which parts do you still consider of "RFC" quality? I'll do one final review when you think "RFC" can be removed (I don't have anything to object to at this point)

@jhedberg RFC removed. I've also fixed the last BT_DEV_READY comment. Thanks for the review.

jhedberg commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -540,41 +584,54 @@ static int bt_ipm_open(void)
+	int err;
+	struct net_buf *rsp;
+
+	err = bt_hci_cmd_send_sync(ACI_HAL_STACK_RESET, NULL, &rsp);
+
+	net_buf_unref(rsp);
+
+	if (err) {
+		BT_ERR("IPM Channel Close Issue");
+		return err;
+	}
```
If `bt_hci_cmd_send_sync()` returns an error, the `rsp` buffer will be undefined (not set), so the unconditional `net_buf_unref(rsp)` is wrong. You might want to check if you have similar faulty patterns elsewhere in the driver.

erwango replied:
Fixed, thanks. Other usages of `bt_hci_cmd_send_sync()` were actually correct.

erwango commented:
@ceolin Would you mind re-re-approving ?

jhedberg commented on drivers/bluetooth/hci/ipm_stm32wb.c:
```diff
@@ -65,6 +65,7 @@ struct aci_set_ble_addr {
 
 #define ACI_WRITE_SET_TX_POWER_LEVEL       BT_OP(BT_OGF_VS, 0xFC0F)
 #define ACI_HAL_WRITE_CONFIG_DATA	   BT_OP(BT_OGF_VS, 0xFC0C)
+#define ACI_HAL_STACK_RESET		   BT_OP(BT_OGF_VS, 0xFC3B)
```
@erwango btw, I noticed that `drivers/bluetooth/hci/spi.c` also has these defines:https://github.com/zephyrproject-rtos/zephyr/blob/3cbfc505b357f80b6aca4cd2dc39c0f60a24c1b1/drivers/bluetooth/hci/spi.c#L106-L109

Perhaps they should be moved into a common header file?

It also seems like the defines here in `ipm_stm32wb.c` have slightly different values - I think the `0xFC..` part of the OCF is redundant, since that'll come from the OGF value when combined with it.
